### PR TITLE
Chinese Tokenization

### PR DESCRIPTION
**Pull request by mnunberg**

### CODE REVIEW

File: cndict/Makefile [hunk 1]
```diff
@@ -0,0 +1,13 @@
+FRISO_INI 		:= friso.ini
+FRISO_LEXDIR	:= lex
+
+all: generated/libcndict_data.a
+
+generated/libcndict_data.a: generated/cndict_data.o
+	$(AR) rcs $@ $^
+
+generated/cndict_data.c: bundle_friso.py
+	python bundle_friso.py -i $(FRISO_INI) -d $(FRISO_LEXDIR) -o generated
```
Comment (dvirsky) [inline cndict/Makefile]: maybe use $(PYTHON) and allow the user to inject it?

File: cndict/bundle_friso.py [hunk 2]
```diff
@@ -0,0 +1,307 @@
+#!/usr/bin/env python
```
Comment (dvirsky) [inline cndict/bundle_friso.py]: Add an explanation of what this script does and why it's here

File: cndict/bundle_friso.py [hunk 3]
```diff
@@ -0,0 +1,307 @@
...
+except OSError as e:
+    if e.errno != errno.EEXIST:
+        raise
+
+dstfile = os.path.join(dstdir, dstfile)
+ofp = open(dstfile, 'w')
+
+if opts.mode == 'c':
+    ofp.write(r'''
+// Generated by python script
```
Comment (dvirsky) [inline cndict/bundle_friso.py]: Be a bit more verbose, what's this file, and why was it generated by a python script, which python script etc...

File: src/document.c [hunk 4]
```diff
@@ -37,6 +39,37 @@ static void freeDocumentContext(void *p) {
...
+}
+static void tokenizerFree(void *p) {
+  RSTokenizer *t = p;
+  t->Free(t);
+}
+
+static RSTokenizer *getTokenizer(const char *language, Stemmer *stemmer, StopWordList *stopwords) {
+  RSTokenizer *t;
+
+  if (language && strcasecmp(language, "chinese") == 0) {
```
Comment (dvirsky) [inline src/document.c]: I suppose we better make sure this is usually called with NULL to avoid the strcasecmp

File: src/ext/default.c [hunk 5]
```diff
@@ -159,18 +161,89 @@ double DisMaxScorer(RSScoringFunctionCtx *ctx, RSIndexResult *h, RSDocumentMetad
...
+typedef struct {
+  int isCn;
+  union {
+    struct {
+      RSTokenizer *tokenizer;
+      Vector *tokList;
+    } cn;
+    struct sb_stemmer *latin;
+  } data;
+} defaultData;
```
Comment (dvirsky) [inline src/ext/default.c]: should be defaultExpanderCtx

File: src/fragmenter.c [hunk 6]
```diff
@@ -114,28 +107,23 @@ static int tokenizerCallback(void *ctxPtr, const Token *tokInfo) {
...
   fragList->docLen = strlen(doc);
-  bufTokenizerCtx ctx = {.fragList = fragList, .terms = terms, .numTerms = numTerms};
-
-  tokenize(doc, &ctx, tokenizerCallback, stemmer, 0, stopwords, TOKENIZE_NOMODIFY);
-  const Fragment *frags = FragmentList_GetFragments(fragList);
-
-  float totalScore = 0;
-  if (fragList->numFrags) {
-    return;
+  RSTokenizer *tokenizer = NewSimpleTokenizer(stemmer, stopwords, TOKENIZE_NOMODIFY);
```
Comment (dvirsky) [inline src/fragmenter.c]: can't we pool this tokenizer?

Comment (mnunberg) [reply src/fragmenter.c]: We can, but this function is only used in tests

**Comment by dvirsky** [comment]
So nevermind

**Comment by dvirsky** [comment]
我们现在支持中文！

**Comment by mnunberg** [comment]
我假装是中国人，希望人们会认为我很酷

**Comment by dvirsky** [comment]
我会有北京烤鸭

**Pull request merged**